app/src/main/cpp/skyline/kernel/memory.cpp

```diff
@@ -199,113 +337,230 @@ namespace skyline::kernel {
-            auto lowerExtension{*lower};
-            lowerExtension.ptr = chunk.ptr + chunk.size;
-            lowerExtension.size = static_cast<size_t>((lower->ptr + lower->size) - lowerExtension.ptr);
-
-            lower->size = static_cast<size_t>(chunk.ptr - lower->ptr);
-            if (lower->size) {
-                upper = chunks.insert(upper, lowerExtension);
-                chunks.insert(upper, chunk);
-            } else {
-                auto lower2{std::prev(lower)};
-                if (chunk.IsCompatible(*lower2) && lower2->ptr + lower2->size >= chunk.ptr) {
-                    lower2->size = static_cast<size_t>(chunk.ptr + chunk.size - lower2->ptr);
-                    upper = chunks.erase(lower);
-                } else {
-                    *lower = chunk;
-                }
-                upper = chunks.insert(upper, lowerExtension);
-            }
-        } else if (chunk.IsCompatible(*lower) && lower->ptr + lower->size >= chunk.ptr) {
-            lower->size = static_cast<size_t>(chunk.ptr + chunk.size - lower->ptr);
-        } else {
-            if (lower->ptr + lower->size > chunk.ptr)
-                lower->size = static_cast<size_t>(chunk.ptr - lower->ptr);
-            if (upper != chunks.end() && chunk.IsCompatible(*upper) && chunk.ptr + chunk.size >= upper->ptr) {
-                upper->ptr = chunk.ptr;
-                upper->size = chunk.size + upper->size;
-            } else {
-                chunks.insert(upper, chunk);
-            }
-        }
+        ForeachChunkinRange(memory, [&](std::pair<u8 *, ChunkDescriptor> &desc) __attribute__((always_inline)) {
+            desc.second.permission = permission;
+            MapInternal(&desc);
+        });
     }
 
-    std::optional<ChunkDescriptor> MemoryManager::Get(void *ptr) {
+    std::optional<std::pair<u8 *, ChunkDescriptor>> MemoryManager::GetChunk(u8 *addr) {
         std::shared_lock lock(mutex);
 
-        auto chunk{std::upper_bound(chunks.begin(), chunks.end(), reinterpret_cast<u8 *>(ptr), [](const u8 *ptr, const ChunkDescriptor &chunk) -> bool { return ptr < chunk.ptr; })};
-        if (chunk-- != chunks.begin())
-            if ((chunk->ptr + chunk->size) > ptr)
-                return std::make_optional(*chunk);
+        if (!addressSpace.contains(addr)) [[unlikely]]
+            return std::nullopt;
+
+        auto chunkBase = upper_bound(addr);
+        if (addr < chunkBase->first) [[likely]]
+            --chunkBase;
+
+        return std::make_optional(*chunkBase);
+    }
+
+    __attribute__((always_inline)) void MemoryManager::MapCodeMemory(span<u8> memory, memory::Permission permission) {
+        std::unique_lock lock(mutex);
+
+        std::pair<u8 *, ChunkDescriptor> temp(
+            memory.data(),
+            ChunkDescriptor{
+                .size = memory.size(),
+                .permission = permission,
+                .state = memory::states::Code});
```

Comment:
move end bracket onto next line, you can put this into MapInternal params directly rather than creating a temp and passing a ptr